glusterd, libglusterfs, api: use arrays for host and peer names

dmantipov opened the pull request:
Introduce `GF_HOST_NAME_MAX` and use fixed-size arrays to manage
host and peer names in `glusterfs_ctx_t` and through `glusterd`
code. This follows the convention to use `PATH_MAX`-sized arrays to
manage file names and eliminates a lot of runtime checks whether
functions like `gf_strdup()` has failed to allocate memory.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

dmantipov commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t

2 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
https://build.gluster.org/job/gh_centos7-regression/2309/

dmantipov commented:
/run regression

dmantipov commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/glusterd/brick-order-check-add-brick.t

0 test(s) generated core 


5 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/georep-basic-dr-rsync.t
./tests/basic/distribute/phase1-migration-fops.t
./tests/bugs/glusterd/brick-order-check-add-brick.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/2316/

dmantipov commented:
/run regression

Commit:
glusterd, libglusterfs, api: use arrays for host and peer names

Introduce 'GF_HOST_NAME_MAX' and use fixed-size arrays to manage
host and peer names in 'glusterfs_ctx_t' and through 'glusterd'
code. This follows the convention to use PATH_MAX-sized arrays to
manage file names and eliminates a lot of runtime checks whether
functions like 'gf_strdup()' has failed to allocate memory.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/glusterfs.h b/libglusterfs/src/glusterfs/glusterfs.h
index 21da033ff..2c1258857 100644
--- a/libglusterfs/src/glusterfs/glusterfs.h
+++ b/libglusterfs/src/glusterfs/glusterfs.h
@@ -63,7 +63,6 @@
 #define O_PATH 010000000 /* from asm-generic/fcntl.h */
 #endif
 
-
 #ifndef EBADFD
 /* Mac OS X does not have EBADFD */
 #define EBADFD EBADF
diff --git a/xlators/mgmt/glusterd/src/glusterd-utils.c b/xlators/mgmt/glusterd/src/glusterd-utils.c
index a4f1a15a3..ec7cfd464 100644
--- a/xlators/mgmt/glusterd/src/glusterd-utils.c
+++ b/xlators/mgmt/glusterd/src/glusterd-utils.c
@@ -235,7 +235,8 @@ glusterd_hostname_new(xlator_t *this, const char *hostname)
         goto out;
     }
 
-    gf_strncpy(hostname_obj->hostname, hostname, sizeof(hostname_obj->hostname));
+    gf_strncpy(hostname_obj->hostname, hostname,
+               sizeof(hostname_obj->hostname));
     CDS_INIT_LIST_HEAD(&hostname_obj->hostname_list);
 out:
     return hostname_obj;
```

xhernandez commented on libglusterfs/src/glusterfs/compat.h:
```diff
@@ -505,6 +505,25 @@ dirname_r(char *path);
+/* Try something generic. */
+#define GF_HOST_NAME_MAX NAME_MAX
+#else /* none of the above */
+#error "Unable to determine maximum host name length"
```
A default of 255, for example, wouldn't be a better option than failing compilation ?

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -4176,11 +4176,8 @@ glusterfs_compute_sha256(const unsigned char *content, size_t size,
+   even if there is no null byte among the bytes copied from 'src'. */
 
 char *
 gf_strncpy(char *dest, const char *src, const size_t dest_size)
```
Probably not for this patch, but I think it would be better to use `snprintf()` or even `memcpy()` to implement this because `strncpy()` always writes the full buffer padding with nulls, even if the source string is much shorter.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-handler.c:
```diff
@@ -164,12 +164,12 @@ glusterd_handle_friend_req(rpcsvc_request_t *req, uuid_t uuid, char *hostname,
     }
 
-    event->peername = gf_strdup(peerinfo->hostname);
+    gf_strncpy(event->peername, peerinfo->hostname, sizeof(event->peername));
```
If for some reason the source hostname is larger than `GF_HOST_NAME_MAX`, shouldn't we fail with an error instead of silently truncating the hostname, which could have unwanted side effects ?
